pkg/zfs/volume.go

```diff
@@ -441,3 +449,79 @@ func IsVolumeReady(vol *apis.ZFSVolume) bool {
 
 	return false
 }
+
+// GetSnapshotForVolume fetches all the snapshots for the given volume
+func GetSnapshotForVolume(volumeID string) (*apis.ZFSSnapshotList, error) {
+	listOptions := metav1.ListOptions{
+		LabelSelector: ZFSVolKey + "=" + volumeID,
+	}
+	snapList, err := snapbuilder.NewKubeclient().WithNamespace(OpenEBSNamespace).List(listOptions)
+	return snapList, err
+}
+
+// MarkForDeletion marks the volume for deletion by adding the annotation
+func MarkForDeletion(volumeName string) error {
+	zv, err := GetZFSVolume(volumeName)
+	if err != nil {
+		klog.Errorf("failed to get ZV %s: %v", volumeName, err)
+		return err
+	}
+
+	err = UpdateZFSVolumeAnnotation(zv)
+	if err != nil {
+		klog.Errorf("failed to update ZV %s: %v", volumeName, err)
+		return err
+	}
+	return nil
+}
+
+// IsVolumeEligibleForDeletion checks if the volume can be deleted or not
+func IsVolumeEligibleForDeletion(volumeName string) (bool, error) {
+	cfg, err := k8sapi.Config().Get()
+	if err != nil {
+		return false, status.Errorf(
+			codes.Internal,
+			"failed to build kubeconfig: %v",
+			err,
+		)
+	}
+
+	clientset, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return false, status.Errorf(
+			codes.Internal,
+			"failed to build k8s clientset: %v",
+			err,
+		)
+	}
+
+	pv, err := clientset.CoreV1().PersistentVolumes().Get(context.TODO(), volumeName, metav1.GetOptions{})
+	// If the PV is present
+	if err == nil {
+		// If the PV exist and does not have the claim reference
+		// then assuming the PVC is not present and thus the ZV is eligible for deletion
+		if pv.Spec.ClaimRef == nil {
+			return true, nil
+		}
+
+		// If the PV exist and has the reclaim policy as retain
+		// then the ZV is not eligible for deletion
+		if pv.Spec.PersistentVolumeReclaimPolicy == "Retain" {
+			return false, nil
```

Comment:
Probably cleanup the annotation if it exists as well? cc @tiagolobocastro ?

Reply:
done